### PR TITLE
Remove deprecated args from `mx.Trainer`; rework LRS.

### PR DESCRIPTION
**Pull request by jaheba**
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Please tag this pr with at least one of these labels to make our release process faster:** BREAKING, new feature, bug fix, other change, dev setup

### CODE REVIEW

File: src/gluonts/mx/trainer/learning_rate_scheduler.py [hunk 1]
```diff
@@ -215,32 +237,4 @@ def on_epoch_end(
...
-        should_continue = self.lr_scheduler.step(metric_value=epoch_loss)
-        if not should_continue:
-            print(
-                "Early stopping based on learning rate scheduler callback"
-                " (min_lr was reached)."
-            )
-            return False
-
-        pre_step_learning_rate = trainer.learning_rate
-        trainer.optimizer.set_learning_rate(
```
Comment (jaheba) [inline src/gluonts/mx/trainer/learning_rate_scheduler.py:227]: This should never alter `trainer.learning_rate`:


```py
In [3]: mx.optimizer.Adam.set_learning_rate??
Signature: mx.optimizer.Adam.set_learning_rate(self, lr)
Source:   
    def set_learning_rate(self, lr):
        """Sets a new learning rate of the optimizer.

        Parameters
        ----------
        lr : float
            The new learning rate of the optimizer.
        """
        if self.lr_scheduler is not None: # pylint: disable=no-else-raise
            raise UserWarning("LRScheduler of the optimizer has already been "
                              "defined. Note that set_learning_rate can mutate "
                              "the value of the learning rate of the optimizer "
                              "only when the LRScheduler of the optimizer is "
                              "undefined.")
        else:
            self.lr = lr
```

File: src/gluonts/mx/trainer/learning_rate_scheduler.py [hunk 2]
```diff
@@ -11,21 +11,55 @@
...
-from gluonts.exceptions import GluonTSUserError
+from gluonts.core.component import validated
 
 from .callback import Callback
 
 
-class MetricAttentiveScheduler(mx.lr_scheduler.LRScheduler):
-    r"""
+@dataclass
+class Metric:
```
Comment (lostella) [inline src/gluonts/mx/trainer/learning_rate_scheduler.py]: Is `Metric` the right name? Seems like the sub-classes do not define actual metrics, but only a strategy for what value to keep track of. Maybe `Objective` is a better term?

File: src/gluonts/mx/trainer/learning_rate_scheduler.py [hunk 3]
```diff
@@ -11,21 +11,55 @@
...
+    best: float
+
+    def update(self, metric: float) -> bool:
+        if self.should_update(metric):
+            self.best = metric
+            return True
+        return False
+
+    def should_update(self, metric: float) -> bool:
+        raise NotImplementedError
```
Comment (lostella) [inline src/gluonts/mx/trainer/learning_rate_scheduler.py:48]: Minor, no real need to change this but: I would put `return True` here

File: test/mx/trainer/test_learning_rate_scheduler.py [hunk 4]
```diff
@@ -113,18 +113,12 @@ def test_PatientMetricAttentiveScheduler(
         # We construct the scheduler with a different rate, because we need
         # to test that the optimizer can override this setting: this is
         # consistent with how schedulers and optimizers operate in MXNet.
-        base_lr=10 * base_lr,
+        learning_rate=base_lr,
```
Comment (lostella) [inline test/mx/trainer/test_learning_rate_scheduler.py]: I guess the comment above does not apply anymore?

Comment (jaheba) [reply test/mx/trainer/test_learning_rate_scheduler.py]: True.

**Review by lostella — APPROVED** [review]

**Comment by lostella** [comment]
@jaheba since this was merged, the following warning started appearing:

```python
[...] src/gluonts/mx/trainer/learning_rate_scheduler.py:67: RuntimeWarning: fields may not start with an underscore, ignoring "_current_patience"
  @dataclass
```

Is this intended, or otherwise worth fixing?

**Comment by lostella** [comment]
sorry, just saw #2166

**Comment by jaheba** [comment]
@lostella I think we should still fix this separately. I think while `dataclasses.dataclass` support underscore, the pydantic version does not.
